Azure/azure-dev
# DevCenter (ADE) components for template source, provision provider & remote environments

Adds the `devcenter` components
- DevCenter template source
- DevCenter provision provider
- DevCenter environment store

## How to test Azure Dev Center Integration

1. Run `azd config set platform.type devcenter`
2. Ensure you you are logged into `azd` with an account that has write access to Dev Center Projects
3. You Dev Center catalogs contain environment definitions with `repoUrl` parameter

When `platform.type` is set to `devcenter` all remote environment state and provisioning will leverage new `devcenter` components.

>[!IMPORTANT]
Only Environment Definitions with `repoUrl` parameter defined will show up as available templates.  
Other environment definitions will show up during `azd provision` if not initialized from the template experience

## How to configure Dev Center in azd

Configurations are merged from a combination of the following:

1. Environment variables
    - `AZURE_DEVCENTER_NAME`
    - `AZURE_DEVCENTER_PROJECT`
    - `AZURE_DEVCENTER_CATALOG`
    - `AZURE_DEVCENTER_ENVIRONMENT_DEFINITION`
    - `AZURE_DEVCENTER_ENVIRONMENT_TYPE`
    - `AZURE_DEVCENTER_ENVIRONMENT_USER`
1. Azd Environment configuration `.azure/<env>/config.json`
    <img width="474" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/b44e4125-1a6f-417c-bad6-f872a598ed87">
1. Azd Project configuration in `azure.yaml`
    <img width="311" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/947755a6-18b1-40db-997d-f9d2d465323f">
1. Azd User configuration in  `~/<user_profile>/.azd/config.json`
    <img width="419" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/72f79921-a681-4edc-b1ee-eb48676e4a74">


> [!NOTE]
`azd` will automatically prompt users for any missing devcenter values

## Effected commands

### `azd init` will prompt to select from valid dev center azd templates

<img width="591" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/bd7815c2-e739-4948-9dfa-c8e43cedfc32">

### `azd template list` will show all available devcenter templates

Diplays a normalized list of available `azd` templates based on dev center projects and their environment definition configuration.  Only includes templates that contain a `repoUrl`

<img width="677" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/ce2bee64-be5c-4a57-9d83-bea33d4940b3">

### `azd provision` will create new devcenter environments

Provision will prompt for any missing value such as dev center environment type.
<img width="416" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/5e09c394-e18e-430c-be9f-d85f0f765afa">

Displays minimal progress in current POC

<img width="403" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/d7a95707-6d71-4df3-a466-66b7219be9b4">

### `azd env list` will show local and remote devcenter environments 

Displays a list of all local and remote dev center environments referencing the same devcenter, project & environment definition

<img width="335" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/3743688e-9f76-472f-88c5-31df6f6c5b33">

Matches the same list you would see in the Developer Portal

<img width="638" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/c10840a0-481a-46f9-95cd-d03abe1580f9">

## `azd up` will continue to package, provision and deploy your application

<img width="734" alt="image" src="https://github.com/Azure/azure-dev/assets/6540159/b9c9d783-e1ad-4037-84e3-2e4644a92ee3">

> [!NOTE]
The `devcentersdk` package is a manual work in progress at the moment till we can incorporate the DevCenter SDK version.
